### ISSUE TITLE
[Social Profiles] Social profiles not saving in metaboxes

### ISSUE THREAD

**Issue by prafgon** [reduxframework/redux-extensions#307]
Hello there!

My purchase ID is: #6530

Social Profiles extension works perfectly fine in the Theme Panel but not in metaboxes. Everything is fine except when saving the post - no changes are remembered. None of the previously selected/added icon is displayed after I save the post which works fine in the theme panel. Can you please take a look at it? I'd truly appreciate it. :)

**UPDATE:** Okay, some super strange things are happening.. after refreshing the post some social icons previously selected appeared. However, after unselecting them, they still appear after save. And when I try to create a new custom post type (with that metabox), the same social profiles are selected as in the post I mentioned previously. Now this is super strange.

Another question: would it be possible to add an option to disable the text/background color pickers? I don't need them and they just slow down the panel. Thanks!

Kind Regards,
FJ

**Comment by kprovance** [comment]
Cant do it today, it's a holiday in America.  I need a Support URL also, as I have to be able to duplicate your issue first and need the specs and config.

**Comment by kprovance** [comment]
I took a quick look at it while waiting for the turkey to cook.  The problem lies mostly with metqboxes, and to a small degree the saving process within social profiles.  Metaboxes is saving your data incorrectly (it's compiled as a URL encoded strong and saved in json format when used via the option panel).  Metaboxes just dumps the url encoded data into the postmeta database, and why changes are not saved, or saved properly.  I'm not sure yet how to make all that work together, as social profiles was never meant to be used in metaboxes.  The metabox extension isn't mine, and the guy who wrote it has stopped working on it, which means at some point Ill have to go into the code and play with it (the unfortunate part is I don't see any proceeds from that extension, it's work I have to do pro bono, and it's going to be days of it).  It might be possible to accomplish something similiar within SP, so the information is correct before metaboxes saves it, but again, that going to take days of trial and error and research, because it has to react to action hook and/or filters from a page/post edit (versus Redux save hooks via the option panel.  Two completely different concepts).  I've got a lot going on IRL for the next three weeks or so.  I'm not sure when I'll be able to sit down and figure it all out.

**Comment by dovy** [comment]
@kprovance Haven't stopped working on it, actually working on an update. Let me know where the issue lies, but you may have to code a hook into your plugin for metaboxes. We'll chat this weekend.  ;)

**Comment by prafgon** [comment]
Oh, I completely forgot you have holidays there. Please don't worry about it and enjoy the Turkey! I hope it didn't overcook because of me - gosh, I wouldn't handle that!

I'm looking forward for your reply after the weekend then, thanks. :)

F.

**Comment by carmoreira** [comment]
Hi! I was also planning to purchase the social profiles extension to use exclusively on metaboxes, on multiple custom post types. Also no hurry, but wanted to leave a comment so you know there are more devs interested in the extension to work with metaboxes also :) Cheers

**Comment by kprovance** [comment]
A goodly amount of work is going to be required.  Maybe if we can get these many debs to commit via crowd funder or some such thing, it'll demonstrate mass interest?  Again, it was never designed to work with metaboxes, and a lot of work will be needed, taking me away from my day job, or having to give up my nights.  A commitment from many devs would be helpful.

**Comment by kprovance** [comment]
Social Profiles 1.0.5 is not compatible with metaboxes.  Log onto your account to download the update (assuming your subscription is current).

@carmoreira - Does this mean I can now expect a mad rush of devs to buy up the extension now?  ;-)

**Comment by prafgon** [comment]
You Sir, are a true hero! Thank you so much. I truly appreciate your effort. :)

It's indeed sad that so few developers actually purchases your products. :(

**Comment by carmoreira** [comment]
Thank you for the update @kprovance ! Just purchased the extension myself! 
Purchase ID#6692. 

It seems it's not working for me yet. Working with version 1.0.5.
The social icons interface display fine in the metaboxes, but as described before by @prafgon it doesn't seem the input is saved correctly. It just resets. I do see something in the database table postmeta, a big array it seems with encoded url info. I can parse that in my frontend, but the problem is the backend, in the metaboxes, where the interface doesn't load the saved data. 

Also, it seems the parameter 'include' is being ignored, since it always displays all icons. Currently using this code, copied from your examples folder:
 
`array(
                'id'            => 'opt-social-profiles',
                'type'          => 'social_profiles',
                'title'         => 'Social Profiles',
                'subtitle'      => 'Click an icon to activate it, drag and drop to change the icon order.',
                'include'       => array(
                    'facebook',
                    'twitter',
                    'yelp'
                ),
            ),`

I can open a separate issue entry for this if you prefer. 

Support URL: http://support.redux.io/?id=zdXpk9zpvSDmPNGhkep%2F0mU3BFIA1TNK9mDeGKhxshsZdhBFCZ5hgb%2FNtziIO0VGscf71vrX4P3orPLPSMCfcQ%3D%3D

Thank you for all the great work developed so far!

**Comment by kprovance** [comment]
Your support URL says you're still using Metaboxes 1.3.0.  I'd try upgrading that first.  Were up to version 1.4.1, I think.  I'll have to get back to you on the include thing.  Compatibility with metaboxes is a bloody NIGHTMARE,  As it was never intended to be, I've ended up having to rewrite huge parts of it.  THis might just be more of the same.

**Comment by kprovance** [comment]
Yeah, the reason includes isnt working in metaboxes, it because of the ID is the same as the one in your option panel, it's going to use the option panel settings, including includes.  Once again, this is going to take some looking into.

**Comment by carmoreira** [comment]
Updated Metaboxes, but the problem persists unfortunately. 
New updated support URL: http://support.redux.io/?id=zdXpk9zpvSDmPNGhkep%2F0mU3BFIA1TNK9mDeGKhxshsZdhBFCZ5hgb%2FNtziIO0VGscf71vrX4P3orPLPSMCfcQ%3D%3D

You're right, somehow it's reading from the options panel - If there's a 'Social Profiles' field in the options panel, then in the metaboxes the field will render the 'include' parameter from the options panel fields array. 

I understand how it can be frustrating to adapt a script to something it wasn't supposed to :\
Will continue testing also and anything I discover I'll post it here.

**Comment by kprovance** [comment]
I have not been able to recreate the saving issue yet.  It works fine on this end.  Ill have to find the time to load up your entire option panel.  The reason SP is pulling the include from the option panel is because that is where the ext looks for the include array.  I'm not even sure yet if I can pull that information from metaboxes.  So far, I have not.  If it means rewriting metaboxes, it's going to be a larger issue.  I don't know yet.  When I do, Ill post back.

**Comment by kprovance** [comment]
Do me a favour, try this:  The database key that has the postmeta data for SP, the one with the url encoded string (which is should not be).  Delete it.  Nuke it clean.  See if that makes a difference.  I think I had to do that on mine before it worked, when I was fixing it.

**Comment by carmoreira** [comment]
It seems the problem is how the data is saved in the postmeta table. It's an array with the info encoded in url type string, like you mentioned in your first reply. The field fails to read the data and considers it's empty. 

Looking at how the extension is reading the data, I tried the following:
- in the file fields_social_profiles.php I changed this block of code:
`if ( is_admin() && ( $pagenow == "post-new.php" || $pagenow == "post.php" ) ) {
                    $post_settings = get_post_meta($post->ID, $this->field_id, true);

                    if (!empty($post_settings)) {
                        $settings = $post_settings;
                    }
                }`
with:

`if ( is_admin() && ( $pagenow == "post-new.php" || $pagenow == "post.php" ) ) {
                    $post_settings = get_post_meta($post->ID, $this->field_id, true);

                    if (!empty($post_settings)) {
                        $settings = array();
                        foreach ($post_settings as $idx => $arr) {
                            foreach ($arr as $key => $value) {
                                $decode = json_decode(urldecode($value),true);
                                array_push($settings,$decode);
                            }
                        }
                        
                    }
                }`


And now the fields save and load correctly in the metaboxes because the $settings array is formatted accordingly. But this is a dirty fix I guess.

**Comment by kprovance** [comment]
That would do it, but it doesn't solve the bigger issue, the settings should not be saving in url encoded format, at all.  That was the 'save_me' function does, it converts it to a proper array, and the before_validate hook in metaboxes is what runs that function and returns the proper data for saving in metaboxes.  Sees that filter isn't running.  Just for giggles look in extension_socil_profiles.php and make sure line 108 says
```
add_filter('redux/metaboxes/save/before_validate', array($this,'save_me'), 0, 3);
```

**Comment by carmoreira** [comment]
Deleting the entries from the database didn't help. 

I confirm line 108:
`            add_filter('redux/metaboxes/save/before_validate', array($this,'save_me'), 0, 3);
`

You're right, the main issue is how the data is being saved. 
Will continue to investigate.

**Comment by kprovance** [comment]
I think at this point, it might save us a TON of time if you just send me your project.  if I can make it fail here, finding the cause will go much faster.  Guessing will take forever.  You can send it to kevin@reduxframework.com

**Comment by carmoreira** [comment]
Sure!

**Comment by kprovance** [comment]
Awesome.  This will go much faster.  I also think I found a way to fix the include issue.  I'm out of town as of Sunday for up to a week(ish), so I'm hoping to get this knocked out now

**Comment by kprovance** [comment]
Did you send it?  I've not gotten it yet

**Comment by carmoreira** [comment]
I did. Maybe double check spam? look for cmoreira.wp 
I am also on slack, I can send it through there.

**Comment by kprovance** [comment]
Got nothing here.  Let's do slack  I'm @conundrum.

**Comment by kprovance** [comment]
Fixed includes and extra icons (same issue).  Also, whomever wanted the option to NOT load color pickers, I added that as well.

**Comment by prafgon** [comment]
Hi again,

I can confirm that Social Profiles extension works perfectly with the latest 1.4.1 version of metaboxes extensions so if profiles do not save for you, please make sure that you have the latest Metaboxes version.

Anyhow, I was the one who asked for a possibility to disable color pickers so kudos for that! However, I can't seem to find an option or parameter to do that? The doc doesn't say anything about it and there is also no IF statement in the extension's source that would check any condition before echoing pickers... The 'include' parameter doesn't seem to be working as expected too so I take a guess a new update is coming up? :)

**Comment by kprovance** [comment]
'color_pickers' => false

You also need the latest version, whatever is posted on the website.

**Comment by prafgon** [comment]
Well, I downloaded it like 40 minutes ago and the latest version is 1.0.5. :)

**Comment by carmoreira** [comment]
Yes, version 1.0.5 still had that issue, which was the problem I had. When social profiles were used in the metaboxes with a unique ID, it didn't save. But Kevin already solved this on version 1.0.7 which I was able to test. The color_pickers argument also works fine on that version. I guess it will be made available soon.
